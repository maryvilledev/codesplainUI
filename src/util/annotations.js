import _ from 'lodash';

export const hasPreviousAnnotation = (annotatedLines, lineNumber) =>
  annotatedLines.includes(lineNumber) && _.head(annotatedLines) !== lineNumber;

export const hasNextAnnotation = (annotatedLines, lineNumber) =>
  annotatedLines.includes(lineNumber) && _.last(annotatedLines) !== lineNumber;

export const getAnnotatedLines = annotations =>
  _.sortBy(_.keys(annotations).map(key => Number(key)));

export const getPreviousAnnotation = (annotations, displayedLineNumber) => {
  const annotatedLines = getAnnotatedLines(annotations);
  if (!hasPreviousAnnotation(annotatedLines, displayedLineNumber)) {
    // First annotation is the one being displayed so there isn't another
    // annotation before this one; return undefined
    return undefined;
  }
  // Get the index of the currently displayed annotation in the annotatedLines
  // array and subtract one to get index of the previous annotation
  const index = _.sortedIndexOf(annotatedLines, displayedLineNumber) - 1;
  return annotations[annotatedLines[index]];
};

export const getNextAnnotation = (annotations, displayedLineNumber) => {
  const annotatedLines = getAnnotatedLines(annotations);
  if (!hasNextAnnotation(annotatedLines, displayedLineNumber)) {
    // Last annotation is the one being displayed so there isn't another
    // annotation after this one; return undefined
    return undefined;
  }
  // Get the index of the currently displayed annotation in the annotatedLines
  // array and add one to get index of the next annotation
  const index = _.sortedIndexOf(annotatedLines, displayedLineNumber) + 1;
  return annotations[annotatedLines[index]];
};
