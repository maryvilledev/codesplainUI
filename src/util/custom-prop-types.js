import { PropTypes } from 'react';

const annotation = PropTypes.shape({
  annotation: PropTypes.string.isRequired,
  lineNumber: PropTypes.number.isRequired,
  lineText: PropTypes.string.isRequired,
});

const annotations = PropTypes.objectOf(annotation);

const filters = PropTypes.shape({
  color: PropTypes.string,
  count: PropTypes.number,
  prettyTokenName: PropTypes.string,
  selected: PropTypes.bool,
});

const lineAnnotated = PropTypes.shape({
  lineNumber: PropTypes.number,
  lineText: PropTypes.string,
});

const permissions = PropTypes.shape({
  canEdit: PropTypes.bool.isRequired,
  canRead: PropTypes.bool.isRequired,
});

const snippetData = PropTypes.shape({
  language: PropTypes.string,
  lastEdited: PropTypes.string,
  snippetTitle: PropTypes.string,
});

const snippets = PropTypes.objectOf(snippetData);

export default {
  annotations,
  filters,
  lineAnnotated,
  permissions,
  snippets,
};
