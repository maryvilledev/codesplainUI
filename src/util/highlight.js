// Default hljs isn't imported as it contains ALL languages (which totals to
// ~520 kb).
import hljs from 'highlight.js/lib/highlight';

// Each language needs to be imported like so:
import bash from 'highlight.js/lib/languages/bash';
import cs from 'highlight.js/lib/languages/cs';
import go from 'highlight.js/lib/languages/go';
import java from 'highlight.js/lib/languages/java';
import javascript from 'highlight.js/lib/languages/javascript';
import php from 'highlight.js/lib/languages/php';
import python from 'highlight.js/lib/languages/python';
import r from 'highlight.js/lib/languages/r';
import ruby from 'highlight.js/lib/languages/ruby';
import x86asm from 'highlight.js/lib/languages/x86asm';

// Register each language to the hljs instance
hljs.registerLanguage('bash', bash);
hljs.registerLanguage('cs', cs);
hljs.registerLanguage('go', go);
hljs.registerLanguage('java', java);
hljs.registerLanguage('javascript', javascript);
hljs.registerLanguage('php', php);
hljs.registerLanguage('python', python);
hljs.registerLanguage('r', r);
hljs.registerLanguage('ruby', ruby);
hljs.registerLanguage('x86asm', x86asm);

export default hljs;
