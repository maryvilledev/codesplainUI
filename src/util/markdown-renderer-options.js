import hljs from 'highlight.js';

export default {
  langPrefix: 'hljs language-',
  highlight: function (str, lang) {
    try {
      if (lang && hljs.getLanguage(lang)) {
        return hljs.highlight(lang, str).value;
      }
      return hljs.highlightAuto(str).value;
    } catch (err) {}

    return ''; // use external default escaping
  }
}
