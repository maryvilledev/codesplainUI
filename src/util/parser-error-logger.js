const {
  NODE_ENV,
  REACT_APP_ENABLE_PARSER_DEBUGGING,
} = process.env;

const isDebuggingEnabled =
  NODE_ENV === 'development' && REACT_APP_ENABLE_PARSER_DEBUGGING === 'true';

const errorFn =
  /* eslint-disable no-console */
  isDebuggingEnabled ? (err) => { console.error(err); } : () => {};

export default errorFn;
