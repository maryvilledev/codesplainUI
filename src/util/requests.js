const API_URL = process.env.REACT_APP_API_URL;

/* Construct the endpoint to make a REST request to. Only the username is
 * field is required; the snippetId will be left blank for POST requests,
 * and can be supplied when needed (say, for PUT requests).
 */
export const makeSaveEndpointUrl = (username, snippetId = '') => {
  if (snippetId) {
    // Return a URL for PUT requests
    return `${API_URL}/users/${username}/snippets/${snippetId}`;
  }
  // Return a URL for POST requests
  return `${API_URL}/users/${username}/snippets`;
}
