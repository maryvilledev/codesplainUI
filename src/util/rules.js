/* Object mapping token types to pretty names and colors for display */
export const rules = {
  // Non-terminal rules
  'and_expr': { prettyName: 'And', color: '#F0E68C' },
  'argument': { prettyName: 'Arguments', color: '#EAF7AB' },
  'arith_expr': { prettyName: 'Arithmetic Expression', color: '#FFA500' },
  'augassign':   { prettyName: 'Augmented Assignment', color: '#00ffff' },
  'break_stmt': { prettyName: 'Break Statement', color: '#00ffa0' },
  'classdef': { prettyName: 'Class Definitions', color: '#03C03C' },
  'comp_op':  { prettyName: 'Comparison Operator', color: '#DAA520' },
  'dictorsetmaker': { prettyName: 'Dictionary', color: '#00FF7F' },
  'except_clause':  { prettyName: 'Except Clauses', color: '#779ECB' },
  'expr':     { prettyName: 'Expressions', color: '#F0ABF7' },
  'for_stmt': { prettyName: 'For Loops', color: '#F7ABAB' },
  'funcdef':  { prettyName: 'Function Definitions', color: '#AEC6CF' },
  'if_stmt':  { prettyName: 'If Statements', color: '#FFFF00' },
  'import_name': { prettyName: 'Import Statement', color: '#FF6666' },
  'integer':  { prettyName: 'Integers', color: '#ABF7C6' },
  'number':   { prettyName: 'Numbers', color: '#DEA5A4' },
  'parameters':  { prettyName: 'Parameters', color: '#FFB347' },
  'pass_stmt':   { prettyName: 'Pass Statements', color: '#FDFD96' },
  'return_stmt': { prettyName: 'Return Statement', color: '#EECCFF' },
  'str':      { prettyName: 'Strings', color: '#CAABF7' },
  'try_stmt': { prettyName: 'Try Statements', color: '#CCCCFF' },
  'while_stmt':  { prettyName: 'While Loops', color: '#F0E68C' },
  // Terminal rules
  '.TRUE': { prettyName: 'Boolean (True)', color: '#FF9933' },
  '.FALSE': { prettyName: 'Boolean (False)', color: '#E67300' },
}

/* Array of rules produced by parser, but ignored by the UI layer */
export const ignoredRules = [
  'atom',
  'suite',
  'file_input',
  'simple_stmt',
  'trailed_atom',
  'trailer',
  'comparison', // Don't ignore after outlines are added to highlighting
  'testlist_comp',
  'arglist',
  'expr_stmt',
  '.IF',
  '.COLON',
  '.NEWLINE',
  '.INDENT',
  '.NAME',
  '.OPEN_PAREN',
  '.STRING_LITERAL',
  '.CLOSE_PAREN',
  '.DEDENT',
  '._EOF',
  '.FOR',
  '.IN',
  '.DECIMAL_INTEGER',
  '.COMMA',
]

/*
Given an AST getRuleCount() recursively populates the specified map with
each token type's occurence count in the AST.
*/
export function getRuleCount(node, map) {
  if (node.type === undefined) return;
  if (map[node.type] === undefined)
    map[node.type] = 1;
  else
    map[node.type] += 1;
  node.children.forEach(child => getRuleCount(child, map));
}

export const generateFilters = (prevFilters, ruleCounts) => {
  const newFilters = {};
  if (!ruleCounts || Object.keys(ruleCounts) === 0) {
    return newFilters;
  }
  Object.keys(ruleCounts)
    .filter(r => rules[r] !== undefined)
    .forEach(r => {
      const selected = prevFilters[r] ? prevFilters[r].selected : false;
      newFilters[r] = {
        prettyTokenName: rules[r].prettyName,
        count:           ruleCounts[r],
        selected:        selected,
        color:           rules[r].color,
      }
    });
  return newFilters;
}
