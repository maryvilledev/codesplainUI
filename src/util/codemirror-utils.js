import { rules, ignoredRules } from './rules.js';
import _ from 'lodash'

/*
Recursive function for highlighting code in a CodeMirror. highlight() is an
exported wrapper func for this, and starts the recursion.
*/
export function highlightNode(codeMirror, node, filters, parentColor) {
  let color = parentColor;

  // If we aren't ignoring this token...
  if (ignoredRules.indexOf(node.type) === -1) {
    const rule = rules[node.type]; // Get the rule obj for this rule
    if (!rule) {
      return; // Remove this return and the highlighting will sometimes fail
    }
    // Use this node's color if it has one
    if (rule.color) {
      color = rule.color;
    }

    // If this token's filter is not selected
    if (!filters[node.type] || !filters[node.type].selected) {
      color = parentColor;
    }

    // Apply the background color CSS to this token
    styleRegion(
      codeMirror,
      node.begin,
      node.end,
      `background-color: ${color};`
    );
  }

  // Highlight all children of this token
  node.children.forEach(child => {
    if (child === Object(child))
      highlightNode(codeMirror, child, filters, color);
  });
}

/*
Given a CodeMirror instance, styleRegion() will apply the specified css style to
the given region of code. The code is treated as a single string, and characters
in that string must be identified by their index (as opposed to row/col). Both
start and end are inclusive.
*/
export function styleRegion(codeMirror, start, end, css) {
  if (end < start) throw new Error('end must be greater than start');
  codeMirror.markText(codeMirror.posFromIndex(start), codeMirror.posFromIndex(end), {css});
}
/*
Given a CodeMirror instance styleLine() will apply the specified css style to the
specified line of code in the editor. The first line is considered line 0, not 1.
*/
export function styleLine(codeMirror, line, css) {
    const lineStart = { line: line, ch: 0 };
    const lineEnd = { line: line, ch: codeMirror.getLine(line).length };
    codeMirror.markText(lineStart, lineEnd, {css: css});
}

/*
Given a CodeMirror instance, styleAll() will apply the specified css style to all
code in the editor.
*/
export function styleAll(codeMirror, css) {
  styleRegion(codeMirror, 0, codeMirror.getValue().length, css);
}

/*
Given a CodeMirror instance, highlight() will use the specified AST and filters
objects to apply highlighting to the code in the CodeMirror editor.
*/
export async function highlight(codeMirror, AST, filters) {
  //Make this a first-class function
  const func = () => highlightNode(codeMirror, AST, filters, 'transparent');
  //Codemirror buffers its calls ahead of time, then performs them atomically
  codeMirror.operation(func)
}
