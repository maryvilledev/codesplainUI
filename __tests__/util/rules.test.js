import * as rules from '../../src/util/rules';

describe('util: rules', () => {
  describe('getRuleCount()', () => {
    it('should return if node.type === undefined', () => {
      const node = {
        type: undefined,
      };
      const ruleCounts = {};
      rules.getRuleCount(node, ruleCounts);
      expect(ruleCounts).toEqual({});
    });

    it('sets uninitialized key values to 1', () => {
      const nodeType = 'gromflomite';
      const node = {
        type: nodeType,
        children: [],
      };
      const ruleCounts = {};
      expect(ruleCounts[nodeType]).toEqual(undefined);
      rules.getRuleCount(node, ruleCounts);
      expect(ruleCounts[nodeType]).toEqual(1);
    });

    it('adds 1 if key value has been initialized', () => {
      const nodeType = 'gromflomite';
      const node = {
        type: nodeType,
        children: [],
      };
      const ruleCounts = {
        [nodeType]: 1,
      };
      rules.getRuleCount(node, ruleCounts);
      expect(ruleCounts[nodeType]).toEqual(2);
    });
    it('recursively applied to the children', () => {
      const nodeType = 'gromflomite';
      const node = {
        type: nodeType,
        children: [
          { type: nodeType, children: [] },
          { type: nodeType, children: [] },
        ],
      };
      const ruleCounts = {};
      rules.getRuleCount(node, ruleCounts);
      expect(ruleCounts[nodeType]).toEqual(3);
    });
  });
  describe('parseCSV()', () => {
    it('parses CSV to rows and columns', () => {
      const csv = `NAME,ROLE,CATCHPHRASE,
Rick,Genius,Wubba Lubba Dub Dub
Morty,Teenager,Aw Geez Rick
`;
      const expected = [
        ['Rick', 'Genius', 'Wubba Lubba Dub Dub'],
        ['Morty', 'Teenager', 'Aw Geez Rick'],
      ];
      expect(rules.parseCSV(csv)).toEqual(expected);
    });
  });
});
