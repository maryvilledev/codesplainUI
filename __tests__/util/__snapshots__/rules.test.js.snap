exports[`util: rules ignored rules should match snapshot 1`] = `
Array [
  "atom",
  "suite",
  "file_input",
  "simple_stmt",
  "trailed_atom",
  "trailer",
  "comparison",
  "testlist_comp",
  "arglist",
  "expr_stmt",
  ".IF",
  ".COLON",
  ".NEWLINE",
  ".INDENT",
  ".NAME",
  ".OPEN_PAREN",
  ".STRING_LITERAL",
  ".CLOSE_PAREN",
  ".DEDENT",
  "._EOF",
  ".FOR",
  ".IN",
  ".DECIMAL_INTEGER",
  ".COMMA",
]
`;

exports[`util: rules rules should match snapshot 1`] = `
Object {
  ".FALSE": Object {
    "color": "#E67300",
    "prettyName": "Boolean (False)",
  },
  ".TRUE": Object {
    "color": "#FF9933",
    "prettyName": "Boolean (True)",
  },
  "and_expr": Object {
    "color": "#F0E68C",
    "prettyName": "And",
  },
  "argument": Object {
    "color": "#EAF7AB",
    "prettyName": "Arguments",
  },
  "arith_expr": Object {
    "color": "#FFA500",
    "prettyName": "Arithmetic Expression",
  },
  "augassign": Object {
    "color": "#00ffff",
    "prettyName": "Augmented Assignment",
  },
  "break_stmt": Object {
    "color": "#00ffa0",
    "prettyName": "Break Statement",
  },
  "classdef": Object {
    "color": "#03C03C",
    "prettyName": "Class Definitions",
  },
  "comp_op": Object {
    "color": "#DAA520",
    "prettyName": "Comparison Operator",
  },
  "dictorsetmaker": Object {
    "color": "#00FF7F",
    "prettyName": "Dictionary",
  },
  "except_clause": Object {
    "color": "#779ECB",
    "prettyName": "Except Clauses",
  },
  "expr": Object {
    "color": "#F0ABF7",
    "prettyName": "Expressions",
  },
  "for_stmt": Object {
    "color": "#F7ABAB",
    "prettyName": "For Loops",
  },
  "funcdef": Object {
    "color": "#AEC6CF",
    "prettyName": "Function Definitions",
  },
  "if_stmt": Object {
    "color": "#FFFF00",
    "prettyName": "If Statements",
  },
  "import_name": Object {
    "color": "#FF6666",
    "prettyName": "Import Statement",
  },
  "integer": Object {
    "color": "#ABF7C6",
    "prettyName": "Integers",
  },
  "number": Object {
    "color": "#DEA5A4",
    "prettyName": "Numbers",
  },
  "parameters": Object {
    "color": "#FFB347",
    "prettyName": "Parameters",
  },
  "pass_stmt": Object {
    "color": "#FDFD96",
    "prettyName": "Pass Statements",
  },
  "return_stmt": Object {
    "color": "#EECCFF",
    "prettyName": "Return Statement",
  },
  "str": Object {
    "color": "#CAABF7",
    "prettyName": "Strings",
  },
  "try_stmt": Object {
    "color": "#CCCCFF",
    "prettyName": "Try Statements",
  },
  "while_stmt": Object {
    "color": "#F0E68C",
    "prettyName": "While Loops",
  },
}
`;
